Added access tokens to GraphQL request headers, updated middleware to handle authorization

jupiter007 opened the pull request:
Issue: #55 

- Added JWT token to headers for graphql requests on both the client- and server-side components. Updated `lib/graphql/client.ts` and `lib/graphql/apollo-wrapper.tsx`
- Added an example client-side component called 'client-graphql-test' to show how apollo client hooks can work on client side using the 'apollo-wrapper'.
- Updated middleware.ts for authorization and redirection and updated middleware spec
- Added a 'get-token' endpoint to get the token from the cookie and added unit test
- Updated auth.ts to use 'jose' instead of 'jsonwebtoken' for token verification due to errors related to the edge runtime not support Node.js 'crypto' module.
- Added utils:
--  `utils/authLink.ts` - Creates 'authLink' for the creation of the client instances
--  `utils/cookiesUtil.ts` - Contains a `deleteCookie` method
--  `utils/getAuthTokenServer.ts` - Gets JWT token from cookie

**To Test authentication and authorization**:
- Run development branch of `dmsp_backend_prototype`. (Make sure the JWT secret is the same for backend as frontend)
- Run this frontend branch and go to `/signup`. Signup with a new user. 
- Go to `/login` and sign in with new user. Notice the new '`dmspt`' httpOnly cookie that was added
- Navigate to `/dmps/10.48321/D1SP4H`. There are currently errors getting data for the page, but you should be able to acces that page.
- Delete the '`dmspt`' cookie and try navigating to `/dmps/10.48321/D1SP4H`. You should be redirected to /login page

**To Test Token in Header for Client Side**:
- Navigate to `/client-graphql-test` and look at your Network tab and filter for graphql. You should see the token in the Authorization header

jupiter007 commented on lib/server/auth.ts:
```diff
@@ -1,23 +1,31 @@
-    // Secret is hardcoded for now
-    // TODO: Need to figure out where the secret will be stored
-    const secret = Buffer.from('Zn8Q5tyZ/G1MHltc4F/gTkVJMlrbKiZt', 'base64');
+/**
```
Changed to using 'jose' package to verify the token. This is because the "Edge runtime environment does not include the 'crypto' Node.js core module, and jsonwebtoken relies on this module for cryptographic operations.

The 'jose' package is designed to work without this, using standard APIs like Web Cryptography API which is supported in browsers.

andrewebdev commented on __tests__/middleware.spec.ts:
```diff
@@ -5,101 +5,94 @@
+        const response = await middleware(request);
 
-        expect(redirectSpy).toHaveBeenCalledWith(`${process.env.NEXT_PUBLIC_BASE_URL}/login`);
+        expect(redirectSpy).toHaveBeenCalledWith('http://localhost:3000/login');
```
Just wondering if there is a reason to hard-code the localhost url here? Keeping it as an environment variable might be preferable, especially since every dev might have their environment set up differently.

andrewebdev replied:
Ah, reading further into the code, looks like all of this is inside the context of a test and mocking the request/response. If this is the case then it's understandable to use a fixed url.

andrewebdev commented on __tests__/middleware.spec.ts:
```diff
@@ -5,101 +5,94 @@
-        expect(request.cookies.delete).toHaveBeenCalledWith('dmspt');
-        expect(redirectSpy).toHaveBeenCalledWith(`${process.env.NEXT_PUBLIC_BASE_URL}/login`);
+        expect(deleteCookie).toHaveBeenCalled();
+        expect(redirectSpy).toHaveBeenCalledWith('http://localhost:3000/login');
```
Same as previous comment on hard-coded url. There are more in this document, but I'm not going to flag each individual one :)

jupiter007 replied:
Yes, @andrewebdev as you mentioned above, I mocked the request and passed in the URL so that I could check the redirect url.

andrewebdev commented on app/api/get-token/__tests__/route.spec.ts:
```diff
@@ -0,0 +1,72 @@
+        (getAuthTokenServer as jest.Mock).mockRejectedValue(new Error('Test error'));
+
+        const response = await GET();
+        expect(response.status).toEqual(405);
```
`405: Method not Allowed` doesn't seem to be the correct error code for this case. The test doesn't seem to check if the correct method (GET, POST etc.) is used.

Further down, the error message also reads as a `Internal Server Error` which should be a `500` code. So it feels like things aren't matching up here.

jupiter007 replied:
Thanks @andrewebdev. The 500 error code totally makes sense to me since it's on the server side. I think I was copying and pasting and forgot to change that one.

andrewebdev commented on app/api/get-token/route.ts:
```diff
@@ -0,0 +1,26 @@
+
+    } catch (err) {
+        console.error('Error fetching token:', err);
+        return NextResponse.json({ error: 'Internal Server Error' }, { status: 405 });
```
From the use case here it seems like `500` should be the correct code to use. 

What would be the cases where a token cannot be found? We probably need to handle each of those cases with different error responses.

If the token cannot be found cause a fault on our side, then `500` would be the correct code. But if it cannot be found cause the user isn't signed in, then `401` is probably the correct code.

But most likely the status `405` here is not the right one, since the method is technically allowed, so long as you have a token.

jupiter007 replied:
Thanks @andrewebdev. I will change it to statusCode of 500.

andrewebdev commented on lib/server/auth.ts:
```diff
@@ -1,23 +1,31 @@
+ * @returns 
+ */
+export function getJwtSecretKey(): Uint8Array {
+    const secret = process.env.JWT_SECRET
```
_Just a side note:_ We should probably set up a mechanism somewhere to easily rotate  secret keys if needed. It's good practice to rotate them every so often to improve site security.

@briri , we should probably create an issue for this so we are reminded of it at a later stage?

briri replied:
My thought was to store this and probably other variables we have yet to identify in AWS Secrets manager which you can configure to rotate on a schedule. 

@jupiter007 can we move this `process.env.JWT_SECRET` to its own function for now? That will make it easier to update when we do swap in Secrets Manager.

jupiter007 replied:
Thanks @andrewebdev and @briri. Using AWS Secrets Manager is a great idea. I'll add a placeholder function for now.

jupiter007 replied:
I'll also create a ticket to use Secrets Manager in the NextJS app.

jupiter007 replied:
I created this ticket to implement Secrets Manager on the frontend: https://github.com/CDLUC3/dmsp_frontend_prototype/issues/62

@briri should I create another ticket for setting up AWS Secrets Manager?

briri commented on app/api/get-token/route.ts:
```diff
@@ -0,0 +1,26 @@
+    }
+};
+
+export const POST = () => {
```
these are great. I like that we explicitly block them

briri reviewed (approved):
this all looks good to me

jupiter007 commented:
@andrewebdev and @briri thanks so much for your code reviews. I really value your input. I've made the changes you requested so far. Please let me know if there is anything else you recommend that I update or add. I'm sure we'll discover that we need more changes as we progress with the project as well.

jupiter007 commented:
Hi @briri and @andrewebdev, I consolidated my es lint configs into .eslintrc.json and fixed some linting issues that weren't caught with "npm lint". For now, I added "npx eslint ." to the "lint" script until I can figure out how to capture all directories.

I also updated some package versions.
